### PR TITLE
feat: Rework server:delete to remove consoleLinks, namespace or openshift oauthAuthorizations

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

### What does this PR do?
This PR have:
 
- New flag(--delete-namespace) for server:delete which allows users to delete the namespace after remove Che resources from cluster
- Remove ConsoleLinks from cluster 
- Remove all oauthClientAuthorizations linked to che instance

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-714
https://github.com/eclipse/che/issues/17784

### CODE REVIEW

File: src/api/kube.ts [hunk 1]
```diff
@@ -1304,19 +1304,29 @@ export class KubeHelper {
...
-  async getAllCheCluster(): Promise<any[]> {
+  async getAllCheCluster(): Promise<any[] | undefined> {
     const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
     try {
       const { body } = await customObjectsApi.listClusterCustomObject('org.eclipse.che', 'v1', 'checlusters')
       return body.items ? body.items : []
     } catch (e) {
+      if (e.response.statusCode === 404) {
+        // There is no CRD 'checlusters`
+        return
```
Comment (tolusha) [inline src/api/kube.ts:1328]: Just return `[]`

File: src/api/kube.ts [hunk 2]
```diff
@@ -1304,19 +1304,29 @@ export class KubeHelper {
 
       return crs[0]
     } catch (e) {
+      if (e.response && e.response.statusCode === 404) {
+        throw this.wrapK8sClientError('Custom Resource "checlusters.org.eclipse" not found.')
```
Comment (tolusha) [inline src/api/kube.ts:1308]: Is it better to `return undefined` ?

File: src/api/kube.ts [hunk 3]
```diff
@@ -1404,6 +1414,52 @@ export class KubeHelper {
     }
   }
 
+  async getOAuthClientAuthorizations(name: string): Promise<string[]> {
```
Comment (tolusha) [inline src/api/kube.ts:1417]: `name` -> `clientName`

File: src/api/kube.ts [hunk 4]
```diff
@@ -1404,6 +1414,52 @@ export class KubeHelper {
...
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const { body } = await customObjectsApi.listClusterCustomObject('oauth.openshift.io', 'v1', 'oauthclientauthorizations')
+
+      if (!body.items) {
+        return []
+      }
+      const oauthClientAuthorizations = body.items as any[]
+      return oauthClientAuthorizations.filter(o => o.metadata.name && o.clientName === name)
```
Comment (tolusha) [inline src/api/kube.ts:1426]: Why do we need to check if `o.metadata.name` exists ?

File: src/api/kube.ts [hunk 5]
```diff
@@ -1404,6 +1414,52 @@ export class KubeHelper {
...
+    } catch (e) {
+      throw this.wrapK8sClientError(e)
+    }
+  }
+
+  async consoleLinkExists(name: string): Promise<boolean> {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const { body } = await customObjectsApi.getNamespacedCustomObject('console.openshift.io', 'v1', '' , 'consolelinks', name)
+      return this.compare(body, name)
```
Comment (tolusha) [inline src/api/kube.ts:1447]: We don't need `this.compare` this is kind of legacy.

File: src/api/kube.ts [hunk 6]
```diff
@@ -1404,6 +1414,52 @@ export class KubeHelper {
...
+    } catch {
+      return false
+    }
+  }
+
+  async deleteConsoleLink(name: string) {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const options = new V1DeleteOptions()
+      await customObjectsApi.deleteNamespacedCustomObject('console.openshift.io', 'v1', '' , 'consolelinks', name, options)
```
Comment (tolusha) [inline src/api/kube.ts:1457]: Is it namespaced or cluster scope object?

File: src/commands/server/delete.ts [hunk 7]
```diff
@@ -28,6 +28,10 @@ export default class Delete extends Command {
   static flags = {
     help: flags.help({ char: 'h' }),
     chenamespace: cheNamespace,
+    'delete-namespace': boolean({
+      description: 'Set to true to delete namespace after removing Eclipse Che',
```
Comment (tolusha) [inline src/commands/server/delete.ts:32]: `Indicates that a Eclipse Che namespace will be deleted as well`

File: src/api/kube.ts (hunk 4 above)
Comment (flacatus) [reply src/api/kube.ts:1426]: I get the name of `oauthClientAuthorizations` from `o.metadata.name` to delete it.

File: src/api/kube.ts (hunk 6 above)
Comment (flacatus) [reply src/api/kube.ts:1457]: is cluster scope object... my mistake

File: src/api/kube.ts (hunk 2 above)
Comment (flacatus) [reply src/api/kube.ts]: Apparently is better to return undefined and don't catch all statusCode.

File: .vscode/launch.json [hunk 8]
```diff
@@ -28,7 +28,25 @@
...
+                "che",
+                "--delete-namespace"
+            ],
+            "console": "integratedTerminal",
+            "internalConsoleOptions": "neverOpen",
+            "disableOptimisticBPs": true,
+            "windows": {
+                "program": "${workspaceFolder}/node_modules/jest/bin/jest",
+            }
+        },
```
Comment (mmorhun) [inline .vscode/launch.json:49]: Do we really need it?

File: src/api/kube.ts [hunk 9]
```diff
@@ -1404,6 +1404,52 @@ export class KubeHelper {
...
+    try {
+      const { body } = await customObjectsApi.listClusterCustomObject('oauth.openshift.io', 'v1', 'oauthclientauthorizations')
+
+      if (!body.items) {
+        return []
+      }
+      const oauthClientAuthorizations = body.items as any[]
+      return oauthClientAuthorizations.filter(o => o.metadata.name && o.clientName === clientName)
+    } catch {
+      return []
```
Comment (mmorhun) [inline src/api/kube.ts:1418]: Shouldn't we throw an error here?

File: .vscode/launch.json (hunk 8 above)
Comment (flacatus) [reply .vscode/launch.json:49]: No, I'll remove it.

File: src/api/kube.ts (hunk 9 above)
Comment (flacatus) [reply src/api/kube.ts:1418]: Not necessary. We can throw an error in case if for some reason is not possible to obtain `oauthClientAuthorizations`

Comment (mmorhun) [reply src/api/kube.ts:1418]: I meant a situation when `customObjectsApi.listClusterCustomObject` fails due to network or permission or so error. Is it safe to return empty list in such case?

Comment (flacatus) [reply src/api/kube.ts:1418]: yes, it is. This functions will return `[]` . For example if there is some network or permissions the error will be thrown here: https://github.com/flacatus/chectl/blob/delete_cleanup/src/api/kube.ts#L1422-L1431

**Review by mmorhun — APPROVED** [review]

File: src/api/kube.ts (hunk 9 above)
Comment (AndrienkoAleksandr) [reply src/api/kube.ts:1418]: Maybe we should at least log this error.

**Review by AndrienkoAleksandr — APPROVED** [review]

File: src/api/kube.ts [hunk 10]
```diff
@@ -1303,8 +1303,8 @@ export class KubeHelper {
       }
 
       return crs[0]
-    } catch (e) {
-      throw this.wrapK8sClientError(e)
+    } catch {
+      return
```
Comment (tolusha) [inline src/api/kube.ts:1307]: I think `return undefined` only if we got 404

File: src/api/kube.ts [hunk 11]
```diff
@@ -1316,8 +1316,8 @@ export class KubeHelper {
     try {
       const { body } = await customObjectsApi.listClusterCustomObject('org.eclipse.che', 'v1', 'checlusters')
       return body.items ? body.items : []
-    } catch (e) {
-      throw this.wrapK8sClientError(e)
+    } catch {
+      return []
```
Comment (tolusha) [inline src/api/kube.ts:1320]: I think `return []` only if we got 404

File: src/api/kube.ts [hunk 12]
```diff
@@ -1404,6 +1404,52 @@ export class KubeHelper {
...
+  async getOAuthClientAuthorizations(clientName: string): Promise<string[]> {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const { body } = await customObjectsApi.listClusterCustomObject('oauth.openshift.io', 'v1', 'oauthclientauthorizations')
+
+      if (!body.items) {
+        return []
+      }
+      const oauthClientAuthorizations = body.items as any[]
+      return oauthClientAuthorizations.filter(o => o.metadata.name && o.clientName === clientName)
```
Comment (tolusha) [inline src/api/kube.ts:1416]: pls don't check `o => o.metadata.name` if we don't need it here.
do this check when it is needed (removing the object)

File: src/api/kube.ts [hunk 13]
```diff
@@ -1404,6 +1404,52 @@ export class KubeHelper {
...
+    }
+  }
+
+  async consoleLinkExists(name: string): Promise<boolean> {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      await customObjectsApi.getClusterCustomObject('console.openshift.io', 'v1', 'consolelinks', name)
+      return true
+    } catch {
+      return false
```
Comment (tolusha) [inline src/api/kube.ts:1439]: I think `return false` only if we got 404.
Will we have any issues on k8s then?

File: src/tasks/che.ts [hunk 14]
```diff
@@ -461,6 +475,19 @@ export class CheTasks {
...
   }
 
+  deleteNamespace(flags: any): ReadonlyArray<Listr.ListrTask> {
+    return [{
+      title: `Delete namespace ${flags.chenamespace}`,
+      task: async (task: any) => {
+        const namespaceExist = await this.kube.namespaceExist(flags.chenamespace)
+        if (namespaceExist) {
+          await this.kube.deleteNamespace(flags.chenamespace)
+          task.title = await `${task.title}...OK`
```
Comment (tolusha) [inline src/tasks/che.ts:485]: Anyway print "OK" even if namespace doesn't exist

File: src/api/kube.ts (hunk 13 above)
Comment (flacatus) [reply src/api/kube.ts]: No, let's return false

File: src/api/kube.ts [hunk 15]
```diff
@@ -1404,6 +1412,64 @@ export class KubeHelper {
...
+      }
+      throw this.wrapK8sClientError(e)
+    }
+  }
+
+  async deleteOAuthClientAuthorizations(oAuthClientAuthorizations: any[]): Promise<boolean> {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const options = new V1DeleteOptions()
+      oAuthClientAuthorizations.forEach(e => {
```
Comment (tolusha) [inline src/api/kube.ts:1438]: first `.filter()` then `.forEach`

File: src/tasks/installers/operator.ts [hunk 16]
```diff
@@ -332,6 +332,19 @@ export class OperatorTasks {
...
     return [{
+      title: 'Delete oauthClientAuthorizations',
+      task: async (_ctx: any, task: any) => {
+        const checluster = await kh.getCheCluster(flags.chenamespace)
+        if (checluster && checluster.spec.auth.oAuthClientName) {
+          const oAuthClientAuthorizations = await kh.getOAuthClientAuthorizations(checluster.spec.auth.oAuthClientName)
+          await kh.deleteOAuthClientAuthorizations(oAuthClientAuthorizations)
+          task.title = await `${task.title}...OK`
+        } else {
+          task.title = await `${task.title}...Skipped: No oauthClientAuthorizations found.`
```
Comment (tolusha) [inline src/tasks/installers/operator.ts:343]: Don't print `SKipped` . Just print `OK` anyway.

File: src/tasks/che.ts [hunk 17]
```diff
@@ -461,6 +475,21 @@ export class CheTasks {
...
+        } else {
+          task.title = await `${task.title}...OK`
```
Comment (tolusha) [inline src/tasks/che.ts:487]: Just print `OK` once after `if` block.

File: src/api/kube.ts (hunk 2 above)
Comment (flacatus) [reply src/api/kube.ts]: Is better to return undefined because if we return [] or something else https://github.com/che-incubator/chectl/blob/master/src/tasks/installers/operator.ts#L340 will not work properly

File: src/api/kube.ts [hunk 18]
```diff
@@ -1404,6 +1412,63 @@ export class KubeHelper {
...
+      }
+      throw this.wrapK8sClientError(e)
+    }
+  }
+
+  async deleteOAuthClientAuthorizations(oAuthClientAuthorizations: any[]): Promise<boolean> {
+    const customObjectsApi = KubeHelper.KUBE_CONFIG.makeApiClient(CustomObjectsApi)
+    try {
+      const options = new V1DeleteOptions()
+      oAuthClientAuthorizations.filter((e => e.metadata && e.metadata.name)).forEach(async e => {
```
Comment (tolusha) [inline src/api/kube.ts:1438]: I hesitates if it really waits until all objects are deleted.

File: src/tasks/che.ts [hunk 19]
```diff
@@ -461,6 +475,20 @@ export class CheTasks {
...
   }
 
+  deleteNamespace(flags: any): ReadonlyArray<Listr.ListrTask> {
+    return [{
+      title: `Delete namespace ${flags.chenamespace}`,
+      task: async (task: any) => {
+        const namespaceExist = await this.kube.namespaceExist(flags.chenamespace)
+        if (namespaceExist) {
+          await this.kube.deleteNamespace(flags.chenamespace)
+          task.title = await `${task.title}...OK`
```
Comment (tolusha) [inline src/tasks/che.ts:485]: We can remove this line.

File: src/api/kube.ts [hunk 20]
```diff
@@ -1404,6 +1412,68 @@ export class KubeHelper {
...
+    } catch (e) {
+      if (e.response.statusCode === 404) {
+        // There is no 'oauthclientauthorizations`
+        return []
+      }
+      throw this.wrapK8sClientError(e)
+    }
+  }
+
+  async deleteOAuthClientAuthorizations(oAuthClientAuthorizations: any[]): Promise<boolean> {
```
Comment (tolusha) [inline src/api/kube.ts:1434]: Just return void instead of boolean

File: src/api/kube.ts [hunk 21]
```diff
@@ -1404,6 +1412,68 @@ export class KubeHelper {
...
+    } catch (e) {
+      if (e.response.statusCode === 404) {
+        // There are no consoleLink
+        return false
+      }
+      throw this.wrapK8sClientError(e)
+    }
+  }
+
+  async deleteConsoleLink(name: string) {
```
Comment (tolusha) [inline src/api/kube.ts:1467]: add :void

**Review by tolusha — APPROVED** [review]

**Comment by flacatus** [comment]
I've finished to test new changes in minikube and openshift and seems everything is correct:

Openshift: 
1. Install Eclipse Che with OAuth +  login into Eclipse Che [I create 2 users to create oauthAuthorizations]
2. server:delete, check if any resources exist(consoleLink, oauthAuths...)... work fine
3. Execute again and check if fix Angel issue [Object object] (https://github.com/eclipse/che/issues/17784).. Fixed
4. Install Eclipse Che without OAuth
5. server:delete to check if resources are removed... Work fine.

Minikube:
Basically I did the same steps like for openshift but without oauth and seems everything works fine

**Comment by flacatus** [comment]
[test]
